### PR TITLE
feat(rust): add output.sourceMapFilename core

### PR DESCRIPTION
**Pull request by lippzhang**
## Related issue (if exists)

[#2938](https://github.com/web-infra-dev/rspack/issues/2938)

## Summary
Manage the output path of source map files
https://webpack.js.org/configuration/output/#outputsourcemapfilename

### <samp>🤖 Generated by Copilot at 0cf70ca</samp>

This pull request adds a new option `sourceMapFilename` to the `rspack` configuration file, which allows users to customize the filename of the source map files generated by the `rspack_plugin_devtool` plugin. It implements this option in the `rspack_core`, `rspack_binding_options`, `rspack_plugin_devtool`, and `rspack_testing` crates, and updates the JSON schema and typescript types for the `rspack` configuration file. It also makes some minor code improvements and updates the `package.json` file.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at 0cf70ca</samp>

*  Add a new `source_map_filename` option to the `rspack` configuration file and the `rspack_core` crate to allow users to configure the output filename of the source map files generated by the `rspack_plugin_devtool` plugin ([link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-ef5856a7fec8410383d58cc835932b263f5df692fd8d31fb1ea236182a162b00R145), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-ef5856a7fec8410383d58cc835932b263f5df692fd8d31fb1ea236182a162b00R190), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-75d001733998d80ed823bf765b9791ff28b8612900dd4a1686a54aa25f3155c5R36), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-75d001733998d80ed823bf765b9791ff28b8612900dd4a1686a54aa25f3155c5R83), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-75d001733998d80ed823bf765b9791ff28b8612900dd4a1686a54aa25f3155c5R98), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-75d001733998d80ed823bf765b9791ff28b8612900dd4a1686a54aa25f3155c5L145-R173), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-a3ed34d20f0b15a073bdeaac181d27e0fe74d19b63f7a0eb6cc1868192ef6dd9L12-R14), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-a3ed34d20f0b15a073bdeaac181d27e0fe74d19b63f7a0eb6cc1868192ef6dd9L170-R193), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-8b7ae8eae42b4d9202dbd35407c3338295c42d2b3b863c1a9c179c23502c86c2R243-R244), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-8b7ae8eae42b4d9202dbd35407c3338295c42d2b3b863c1a9c179c23502c86c2R407-R408), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-9a3df8494a717f2b873f823e39ca33e5b177d14e63719631320d675cfc6909eaR408-R411), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L80-L79), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-5f5bb1aeb2a977ff30c5ab822c79f09b76476c503462939c69beb9e4d7552123L191-R192), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-55e0fbd74437dff4e251152f5d93efd7f5871cbad72a7959dddb6add8fbf60ccR497-R499), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-78648f723799beccc94c4b2e506997d12d1ace8216aeb31c0731c37a6dce024aR77), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88R452-R460), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88R1172-R1174), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fR150), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fR168), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fR252))
*  Use the `[file]` placeholder syntax for the `source_map_filename` option to support the original filename in the template, similar to `webpack` ([link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-75d001733998d80ed823bf765b9791ff28b8612900dd4a1686a54aa25f3155c5R83), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-75d001733998d80ed823bf765b9791ff28b8612900dd4a1686a54aa25f3155c5L145-R173))
*  Check the `devtool` option and the `public_path` option in the `rspack_plugin_devtool` plugin to handle different modes and paths for the source map files ([link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-a3ed34d20f0b15a073bdeaac181d27e0fe74d19b63f7a0eb6cc1868192ef6dd9L170-R193))
*  Use the `..Default::default()` syntax for the `AssetInfo` struct initialization in the `rspack_plugin_asset`, `rspack_plugin_copy`, and `rspack_plugin_html` plugins to improve the code readability and maintainability ([link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-60327d77238d48f4c8ce2d4c1415530ee60e71ac5c2b879b608209b83189e5edR293), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-60327d77238d48f4c8ce2d4c1415530ee60e71ac5c2b879b608209b83189e5edR530), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-cde7edc635e72d8f64c982c4e78298d248e671f6f022318cc20440f6091b70baR207), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-a889d0e910b514418c3aefad6b3a13a395a9a030ab135118edcf0d82bd2673fcR203))
*  Fix the indentation of the test case file for the `source-map` feature in the `webpack-test` directory ([link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-f38605e2030e9d1d7e83e2830f0bb851faad7190e948ee41fb90d4204f5ac8aaL1-R1), [link](https://github.com/web-infra-dev/rspack/pull/3005/files?diff=unified&w=0#diff-f38605e2030e9d1d7e83e2830f0bb851faad7190e948ee41fb90d4204f5ac8aaL8-R9))

</details>

### CODE REVIEW

**Comment by ahabhgk** [comment]
General looks good to me, ready to merge after rebase

File: crates/rspack_plugin_devtool/src/lib.rs [hunk 1]
```diff
@@ -167,7 +167,26 @@ impl Plugin for DevtoolPlugin {
...
+          let source_map_filename_config = &args.compilation.options.output.source_map_filename;
+          for chunk in args.compilation.chunk_by_ukey.values() {
+            for file in &chunk.files {
+              if *file.to_string() == filename {
+                let extension = if is_css { ".css" } else { ".js" };
+                source_map_filename = source_map_filename_config.render_with_chunk_and_file(
+                  chunk,
+                  &filename,
+                  extension,
+                  &SourceType::JavaScript,
```
Comment (lippzhang) [inline crates/rspack_plugin_devtool/src/lib.rs:182]: maybe  &SourceType::css ?

Comment (lippzhang) [reply crates/rspack_plugin_devtool/src/lib.rs]: fix

**Comment by lippzhang** [comment]
> General looks good to me, ready to merge after rebase

have rebase main branch

**Review by ahabhgk — APPROVED** [review]
Thanks!

**Pull request merged**